File: platform/graphics/texture.go
package graphics

import (
	"fmt"

	rl "github.com/gen2brain/raylib-go/raylib"
	"github.com/hellory4n/stellarthing/core"
)

// texture :)
type Texture struct {
	// these are used for converting to raylib textures
	id      uint32
	width   int32
	height  int32
	mipmaps int32
	format  rl.PixelFormat
	path    string
}

var textureCache map[string]Texture

func init() {
	textureCache = make(map[string]Texture)
}

// supported formats are PNG, BMP, TGA, JPG, GIF, QOI, PSD, DDS, HDR, KTX, ASTC, PKM, and PVR
func LoadTexture(path string) Texture {
	texture, ok := textureCache[path]
	if !ok {
		var txddrx rl.Texture2D = rl.LoadTexture(core.AssetPath(path))
		texture = Texture{
			id:      txddrx.ID,
			width:   txddrx.Width,
			height:  txddrx.Height,
			mipmaps: txddrx.Mipmaps,
			format:  txddrx.Format,
			path:    path,
		}
		textureCache[path] = texture
		fmt.Printf("[GRAPHICS] Loaded texture at %v\n", path)
	}
	return texture
}

// returns the size of the texture in pixels
func (t Texture) Size() core.Vec2i {
	return core.Vec2i{int64(t.width), int64(t.height)}
}

// frees the texture and removes it from the cache. if you want to reload the texture, use this
func (t Texture) Free() {
	var rltltx rl.Texture2D = rl.Texture2D{
		ID:      t.id,
		Width:   t.width,
		Height:  t.height,
		Mipmaps: t.mipmaps,
		Format:  t.format,
	}
	rl.UnloadTexture(rltltx)
	delete(textureCache, t.path)
	fmt.Printf("[GRAPHICS] Freed texture at %v\n", t.path)
}

func FreeAllTextures() {
	for _, v := range textureCache {
		v.Free()
	}
}

File: platform/graphics/texture.c
#include <raylib.h>
#include "core/hashmap.h"
#include "core/list.h"
#include "texture.h"

StHashMap* textures;

void st_init_textures(void)
{
	textures = StHashMap_new(0);
}

void st_free_textures(void)
{
	// free items in the hashmap
	StList* items = StHashMap_items(textures);
	for (nint i = 0; i < items->length; i++) {
		StTexture_free(((StKeyValue*)items->data[i])->val);
		st_log("[ASSETS] Freed texture at %s", (char*)((StKeyValue*)items->data[i])->key);
	}
	StList_free_all(items);

	StHashMap_free(textures);
}

StTexture* StTexture_new(const char* path)
{
	// loading the same textures 385980320859 times isn't very salubrious
	if (StHashMap_contains(textures, path)) {
		return StHashMap_get(textures, path);
	}

	StTexture* dog = malloc(sizeof(StTexture));
	Texture2D mate = LoadTexture(path);
	dog->id = mate.id;
	dog->width = mate.width;
	dog->height = mate.height;
	dog->mipmaps = mate.mipmaps;
	dog->format = mate.format;
	StHashMap_set(textures, path, dog);

	st_log("[ASSETS] Loaded texture at %s", path);
	return dog;
}

void StTexture_free(StTexture* t)
{
	// OH MY GAMMIT IT'S... tally hall
	Texture2D mmmm_mmmm_mmmm = {
		.id = t->id,
		.width = t->width,
		.height = t->height,
		.format = t->format,
		.mipmaps = t->mipmaps,
	};
	UnloadTexture(mmmm_mmmm_mmmm);
}